File: android/app/src/main/kotlin/dev/andromac/core/Crypto.kt
package dev.andromac.core

import java.security.AlgorithmParameters
import java.security.KeyFactory
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.MessageDigest
import java.security.PrivateKey
import java.security.PublicKey
import java.security.SecureRandom
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec
import java.security.spec.ECParameterSpec
import java.security.spec.ECPoint
import java.security.spec.ECPublicKeySpec
import java.security.spec.PKCS8EncodedKeySpec
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * P-256 ECDH + HKDF-SHA256 + AES-256-GCM. The counterpart of docs/PROTOCOL.md §2 and §4.
 * No third-party dependencies; everything here comes from the platform JCE.
 */
object Crypto {

    const val PUB_LEN = 65          // X9.62 uncompressed: 0x04 || X(32) || Y(32)
    private const val CURVE = "secp256r1"

    private val params: ECParameterSpec by lazy {
        AlgorithmParameters.getInstance("EC").run {
            init(ECGenParameterSpec(CURVE))
            getParameterSpec(ECParameterSpec::class.java)
        }
    }

    fun generateKeyPair(): KeyPair =
        KeyPairGenerator.getInstance("EC").apply { initialize(ECGenParameterSpec(CURVE), SecureRandom()) }
            .generateKeyPair()

    /** Public key -> 65-byte X9.62 uncompressed form. */
    fun encodePublic(key: PublicKey): ByteArray {
        val w = (key as ECPublicKey).w
        val out = ByteArray(PUB_LEN)
        out[0] = 0x04
        fixedWidth(w.affineX.toByteArray()).copyInto(out, 1)
        fixedWidth(w.affineY.toByteArray()).copyInto(out, 33)
        return out
    }

    /** 65-byte X9.62 -> PublicKey. Points that are not on the curve are rejected here. */
    fun decodePublic(bytes: ByteArray): PublicKey {
        require(bytes.size == PUB_LEN && bytes[0] == 0x04.toByte()) { "invalid P-256 public key" }
        val x = java.math.BigInteger(1, bytes.copyOfRange(1, 33))
        val y = java.math.BigInteger(1, bytes.copyOfRange(33, 65))
        // KeyFactory verifies that the point lies on the curve; an off-curve point throws InvalidKeySpecException.
        return KeyFactory.getInstance("EC").generatePublic(ECPublicKeySpec(ECPoint(x, y), params))
    }

    fun encodePrivate(key: PrivateKey): ByteArray = key.encoded            // PKCS#8
    fun decodePrivate(bytes: ByteArray): PrivateKey =
        KeyFactory.getInstance("EC").generatePrivate(PKCS8EncodedKeySpec(bytes))

    fun ecdh(own: PrivateKey, peer: PublicKey): ByteArray =
        KeyAgreement.getInstance("ECDH").run { init(own); doPhase(peer, true); generateSecret() }

    fun sha256(vararg parts: ByteArray): ByteArray =
        MessageDigest.getInstance("SHA-256").apply { parts.forEach(::update) }.digest()

    fun hkdf(salt: ByteArray, ikm: ByteArray, info: String, len: Int): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(salt, "HmacSHA256"))
        val prk = mac.doFinal(ikm)                                          // Extract

        mac.init(SecretKeySpec(prk, "HmacSHA256"))                          // Expand
        val infoBytes = info.toByteArray(Charsets.UTF_8)
        val out = ByteArray(len)
        var t = ByteArray(0)
        var pos = 0
        var counter = 1
        while (pos < len) {
            mac.update(t); mac.update(infoBytes); mac.update(counter.toByte())
            t = mac.doFinal()
            val n = minOf(t.size, len - pos)
            t.copyInto(out, pos, 0, n)
            pos += n; counter++
        }
        return out
    }

    fun nonce(counter: Long): ByteArray {
        val n = ByteArray(12)
        for (i in 0 until 8) n[11 - i] = (counter ushr (8 * i)).toByte()    // 4 zero bytes + BE64 counter
        return n
    }

    fun seal(key: ByteArray, counter: Long, plaintext: ByteArray): ByteArray =
        Cipher.getInstance("AES/GCM/NoPadding").run {
            init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce(counter)))
            doFinal(plaintext)
        }

    fun open(key: ByteArray, counter: Long, ciphertext: ByteArray): ByteArray =
        Cipher.getInstance("AES/GCM/NoPadding").run {
            init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce(counter)))
            doFinal(ciphertext)
        }

    const val NONCE_LEN = 32        // n_A / n_B and the commitment c_B, PROTOCOL §2
    const val TAG_LEN = 16          // AES-GCM tag, appended to every ciphertext

    fun randomBytes(n: Int): ByteArray = ByteArray(n).also(SecureRandom()::nextBytes)

    /** docs/PROTOCOL.md §2 — c_B = SHA256("AndroMac/commit" || n_B). */
    fun commit(nonce: ByteArray): ByteArray = sha256("AndroMac/commit".toByteArray(Charsets.UTF_8), nonce)

    /**
     * docs/PROTOCOL.md §3 — the 6 digits compared across the two screens. Bound to the
     * session transcript and both fresh nonces, so a MITM cannot grind the code offline.
     */
    fun sas(transcript: ByteArray, nA: ByteArray, nB: ByteArray): String {
        val h = sha256("AndroMac/SAS/v3".toByteArray(Charsets.UTF_8), transcript, nA, nB)
        var v = 0L
        for (i in 0 until 4) v = (v shl 8) or (h[i].toLong() and 0xff)
        return (v % 1_000_000L).toString().padStart(6, '0')
    }

    fun constantTimeEquals(a: ByteArray, b: ByteArray): Boolean = MessageDigest.isEqual(a, b)

    /** BigInteger.toByteArray() may prepend a sign byte or come back short; pin it to 32 bytes. */
    private fun fixedWidth(raw: ByteArray): ByteArray = when {
        raw.size == 32 -> raw
        raw.size > 32 -> raw.copyOfRange(raw.size - 32, raw.size)
        else -> ByteArray(32).also { raw.copyInto(it, 32 - raw.size) }
    }
}

File: macos/Sources/AndroMacKit/Crypto.swift
import CryptoKit
import Foundation

/// P-256 ECDH + HKDF-SHA256 + AES-256-GCM. docs/PROTOCOL.md §2 and §4.
/// Must stay byte-for-byte compatible with `Crypto.kt` on the Android side.
public enum Crypto {

    public static let pubLen = 65      // X9.62 uncompressed

    // MARK: keys

    public static func generateKeyPair() -> P256.KeyAgreement.PrivateKey {
        P256.KeyAgreement.PrivateKey()
    }

    public static func encodePublic(_ key: P256.KeyAgreement.PublicKey) -> Data {
        key.x963Representation                       // 0x04 || X(32) || Y(32)
    }

    /// Points that are not on the curve are rejected here (CryptoKit validates them).
    public static func decodePublic(_ bytes: Data) throws -> P256.KeyAgreement.PublicKey {
        guard bytes.count == pubLen else { throw WireError.protocolError("invalid public key length") }
        return try P256.KeyAgreement.PublicKey(x963Representation: bytes)
    }

    public static func ecdh(_ own: P256.KeyAgreement.PrivateKey, _ peer: P256.KeyAgreement.PublicKey) throws -> Data {
        let secret = try own.sharedSecretFromKeyAgreement(with: peer)
        return secret.withUnsafeBytes { Data($0) }
    }

    // MARK: KDF

    public static func sha256(_ parts: Data...) -> Data {
        var h = SHA256()
        for p in parts { h.update(data: p) }
        return Data(h.finalize())
    }

    /// RFC 5869. Produces the same result as the hand-written HKDF on Android.
    public static func hkdf(salt: Data, ikm: Data, info: String, count: Int) -> SymmetricKey {
        HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: ikm),
            salt: salt,
            info: Data(info.utf8),
            outputByteCount: count
        )
    }

    // MARK: AEAD

    /// 12 bytes: 4 zero bytes + a big-endian 64-bit counter.
    public static func nonce(_ counter: UInt64) throws -> AES.GCM.Nonce {
        var raw = Data(count: 4)
        raw.append(contentsOf: withUnsafeBytes(of: counter.bigEndian) { Data($0) })
        return try AES.GCM.Nonce(data: raw)
    }

    /// CAUTION: the output of Java's `Cipher.doFinal` is `ciphertext || tag`; it carries no nonce.
    /// That is why CryptoKit's `.combined` field (which also contains the nonce) is NOT USED.
    public static func seal(_ key: SymmetricKey, _ counter: UInt64, _ plaintext: Data) throws -> Data {
        let box = try AES.GCM.seal(plaintext, using: key, nonce: try nonce(counter))
        return box.ciphertext + box.tag
    }

    public static func open(_ key: SymmetricKey, _ counter: UInt64, _ payload: Data) throws -> Data {
        guard payload.count > 16 else { throw WireError.protocolError("frame too short") }
        let box = try AES.GCM.SealedBox(
            nonce: try nonce(counter),
            ciphertext: payload.prefix(payload.count - 16),
            tag: payload.suffix(16)
        )
        return try AES.GCM.open(box, using: key)
    }

    // MARK: SAS

    public static let nonceLen = 32      // n_A / n_B and the commitment c_B, PROTOCOL §2
    public static let tagLen = 16        // AES-GCM tag, appended to every ciphertext

    public static func randomBytes(_ count: Int) -> Data {
        SymmetricKey(size: .init(bitCount: count * 8)).withUnsafeBytes { Data($0) }
    }

    /// docs/PROTOCOL.md §2 — c_B = SHA256("AndroMac/commit" || n_B).
    public static func commit(_ nonce: Data) -> Data {
        sha256(Data("AndroMac/commit".utf8), nonce)
    }

    /// docs/PROTOCOL.md §3 — the 6 digits compared on the two screens. Bound to the session
    /// transcript and both fresh nonces, so a MITM cannot grind the code offline.
    public static func sas(transcript: Data, nA: Data, nB: Data) -> String {
        let h = sha256(Data("AndroMac/SAS/v3".utf8), transcript, nA, nB)
        let v = h.prefix(4).reduce(UInt64(0)) { ($0 << 8) | UInt64($1) }
        return String(format: "%06d", v % 1_000_000)
    }

    /// Constant-time comparison.
    public static func constantTimeEquals(_ a: Data, _ b: Data) -> Bool {
        guard a.count == b.count else { return false }
        var diff: UInt8 = 0
        for (x, y) in zip(a, b) { diff |= x ^ y }
        return diff == 0
    }
}

public enum WireError: LocalizedError {
    case protocolError(String)
    /// The peer authenticated itself, but its static key is not one we have pinned.
    /// `isFirstDevice` is true when nothing is paired yet — the difference between "set this up"
    /// and "add another phone", which is wording, not security. The SAS is what the user checks.
    case untrusted(peerKey: Data, peerName: String, sas: String, isFirstDevice: Bool)
    case closed

    public var errorDescription: String? {
        switch self {
        case .protocolError(let m): return m
        case .untrusted(_, _, let sas, let mismatch):
            return mismatch ? "the pinned key has changed" : "device is not paired (code \(sas))"
        case .closed: return "connection closed"
        }
    }
}
